Plans on JP: Show free for first year offer on domain selection screen

staskus opened the pull request:
Fixes #20686

Configure Domain Suggestions View Controller for Free to Paid plans flow:
1. Specify domains being free for the first year
2. Present from the dashboard card
3. Create a structure for the flow that can be used for integrating plan selection and checkout

📒 Domain selection does not show any premium domains that would not be eligible to appear for free (premium domains). We could consider improving the domain selection view or integrating a new domain selection view used in site creation as a task outside the project. The current domain selection view is quite basic on iOS and lags behind the user experience on the web.

## To test:

1. Log into free .com account
2. Enable "Free to Paid Plans Dashboard Card" feature flag
3. Tap on the dashboard card
5. Confirm that domains view is opened
6. Confirm that "free for the first year" offer appears

## Regression Notes
1. Potential unintended areas of impact

None

2. What I did to test those areas of impact (or what existing automated tests I relied on)

None

3. What automated tests I added (or what prevented me from doing so)

None

PR submission checklist:

- [x] I have completed the Regression Notes.
- [x] I have considered adding unit tests for my changes.
- [x] I have considered adding accessibility improvements for my changes.
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

https://github.com/wordpress-mobile/WordPress-iOS/assets/4062343/dd784916-649d-46a2-92fc-8869da29b939

staskus committed:
Show Free for First Year indication for domain suggestion cell for mapped domain type

staskus committed:
Configure DomainSuggestionsViewController for free to paid plans flow

- Handle .mapped domain case
- Provide a callback when domain is added to cart

staskus committed:
Create FreeToPaidPlansCoordinator to contain free to paid plans domains flow

staskus committed:
Present domain suggestion from FreeToPaidPlansDashboardCard

guarani commented on WordPress/Classes/ViewRelated/Domains/Domain registration/RegisterDomainSuggestions/RegisterDomainSuggestionsViewController.swift:
```diff
@@ -218,7 +219,16 @@ extension RegisterDomainSuggestionsViewController: NUXButtonViewControllerDelega
+                self?.presentWebViewForCurrentSite(domainSuggestion: domain)
+                self?.setPrimaryButtonLoading(false, afterDelay: 0.25)
+            }
+        case .mapped:
```
I'm not sure I understand why `.mapped` is used here, I expected `.registered` instead. Users are registering a domain with WP.com, so this fits [the description of `.registered`](https://github.com/wordpress-mobile/WordPressKit-iOS/blob/c323bc92ef806679bf830f87383d343f0f2f6469/WordPressKit/RemoteDomain.swift#L11-L12).

I see `.mapped` [described here](https://github.com/wordpress-mobile/WordPressKit-iOS/blob/c323bc92ef806679bf830f87383d343f0f2f6469/WordPressKit/RemoteDomain.swift#L13-L14) as a user bringing a domain from somewhere else and connecting it to their WP.com site and that doesn't seem to fit this flow.

staskus replied:
Good point, I misunderstood `mapped` case.

I think then it's better to configure `RegisterDomainSuggestionViewController` not with `domainType` but by `flowType`, since we may want to have different configurations for same domain types.

staskus replied:
[I made a bunch of renamings](https://github.com/wordpress-mobile/WordPress-iOS/pull/20697/commits/d4c5fa41b927f8900edde861773511180daa223f) to reflect why `domainType` was used in these view controllers: to differentiate between different domain purchasing flows.

Hopefully, it will increase clarity. Clean-code-wise, it could probably be replaced by inheritance rather than `enum` to only have a couple of overridden methods that would differentiate between different cases. However, I don't think this usage of `enums` is too intrusive in this case.

guarani replied:
Great idea with the re-name, it looks good!

guarani reviewed (commented):
> 📒 Domain selection does not show any premium domains that would not be eligible to appear for free (premium domains). We could consider improving the domain selection view or integrating a new domain selection view used in site creation as a task outside the project. The current domain selection view is quite basic on iOS and lags behind the user experience on the web.

An example of the premium domain you're referring to is the one the purple arrow is pointing to below, right?

| <img width="1098" alt="Screenshot 2023-05-18 at 15 54 40" src="https://github.com/wordpress-mobile/WordPress-iOS/assets/1898325/53a4e9e6-ee34-477f-853d-b5be159a2a64"> |
| - |

I've also highlighted in green the domains I see on web and mobile with the default domain search (which was "site title"). There's a couple shown on web that aren't shown on mobile e.g. `sitebook.store` but I haven't dug into what we're sending to the API yet to get these suggestions and what differences there could be on mobile and web.

staskus commented:
> An example of the premium domain you're referring to is the one the purple arrow is pointing to below, right?

Yes. 

> I've also highlighted in green the domains I see on web and mobile with the default domain search (which was "site title"). There's a couple shown on web that aren't shown on mobile e.g. sitebook.store but I haven't dug into what we're sending to the API yet to get these suggestions and what differences there could be on mobile and web.

Exactly, the mobile domain search we use right now definitely could and most probably should be improved. Do you see such improvements as part of the project, @guarani ?

staskus committed:
Define a domain selection type to differentiate between different domain purchasing flows

- Defined DomainSelectionType
- Replaced usage of DomainType to DomainSelectionType to reflect 3 currently supported flows: registering domain with a domain credit, purchasing domain directly, and purchasing domain with a paid plan.

guarani commented:
> Exactly, the mobile domain search we use right now definitely could and most probably should be improved. Do you see such improvements as part of the project, @guarani ?

No, I think it's out of scope. I just made [a draft issue](https://github.com/orgs/wordpress-mobile/projects/154/views/1?pane=issue&itemId=28650005) to collect this and anything else that comes up and is out of scope.

guarani commented on WordPress/Classes/ViewRelated/Domains/Domain registration/RegisterDomainSuggestions/RegisterDomainSuggestionsViewController.swift:
```diff
@@ -12,11 +18,12 @@ class RegisterDomainSuggestionsViewController: UIViewController {
     private var siteName: String?
     private var domainsTableViewController: DomainSuggestionsTableViewController?
-    private var domainType: DomainType = .registered
+    private var domainSelectionType: DomainSelectionType = .purchaseSeparately
```
I think `.registered` corresponds to `.registerWithPaidPlan` (which happens when the user buys a plan and has domain credits they haven't used yet). 
I think `.purchaseSeparately` corresponds to `.siteRedirect` (when the user buys a domain without a plan).

guarani replied:
```suggestion
    private var domainSelectionType: DomainSelectionType = . registerWithPaidPlan
```

staskus replied:
Thanks for spotting. I claimed that this naming made it *easier* but I still confused myself. 🤦‍♂️

guarani commented on WordPress/Classes/ViewRelated/Domains/Domain registration/RegisterDomainSuggestions/RegisterDomainSuggestionsViewController.swift:
```diff
@@ -44,13 +51,13 @@ class RegisterDomainSuggestionsViewController: UIViewController {
 
     static func instance(site: Blog,
-                         domainType: DomainType = .registered,
+                         domainSelectionType: DomainSelectionType = .purchaseSeparately,
```
```suggestion
                         domainSelectionType: DomainSelectionType = . registerWithPaidPlan,
```

guarani reviewed (commented):
Hey @staskus, I think the re-name here makes sense but have a question about one of the mappings.

staskus committed:
Revert to correct default parameter for domain suggestions view controller

guarani reviewed (approved):
Works great! I did a quick smoke test of:
- [x] "Find a custom domain" card: doesn't show "Free for the first year" on domains 
- [x] "Free domain with an annual plan" card: shows "Free for the first year" on domains 
- [x] My Site > Menu on site with domain credits: shows "Free for the first year" on domains

Thanks @staskus!

guarani commented:
Merging now since this doesn't seem to have anything left to do. HTH
